src/measurements.jl

```diff
@@ -0,0 +1,184 @@
+
+# for every sites
+measure(ψ::MPS, measurement::String, s::Vector{<:Index}) = 
+  measure(ψ::MPS, (measurement, 1:length(ψ)),s)
+
+## for every sites
+measure(ψ::MPS, measurement::String) = 
+  measure(ψ::MPS, (measurement, 1:length(ψ)), siteinds(ψ))
+
+
+# at a given site
+"""
+    measure(ψ::MPS, measurement::Tuple{String,Int,String,Int}, s::Vector{<:Index})
+
+
+Perform a measurement of a 2-body tensor-product operator on an MPS ψ. The two operators
+are defined by Strings (for op name) and the sites. If an additional set of indices `s` is provided, the correct site is 
+extracted by comparing the MPS with the index order in `s`.
+"""
+function measure(ψ::MPS, measurement::Tuple{String,Int,String,Int}, s::Vector{<:Index})
+  obsA  = measurement[1]
+  obsB  = measurement[3]
+  siteA0 = measurement[2]
+  siteB0 = measurement[4]
+  siteA = findsite(ψ,s[siteA0])
+  siteB = findsite(ψ,s[siteB0])
+  
+  if siteA > siteB
+    obsA, obsB  = obsB, obsA
+    siteA,siteB = siteB,siteA
+  end
+  ϕ = orthogonalize!(copy(ψ), siteA)
+  ϕdag = prime(dag(ϕ),tags="Link")
+  
+  if siteA == siteB
+    C = ϕ[siteA] * gate(obsA, firstsiteind(ϕ, siteA))
+    C = noprime(C,tags="Site") * gate(obsA, firstsiteind(ϕ, siteA)) 
+    C = noprime(C,tags="Site") * noprime(ϕdag[siteA])
+    return real(C[])
+  end
+  if siteA == 1
+    C = ϕ[siteA] * gate(obsA, firstsiteind(ϕ, siteA))
+    C = noprime(C,tags="Site") * ϕdag[siteA]
+  else
+    C = prime(ϕ[siteA],commonind(ϕ[siteA],ϕ[siteA-1])) * gate(obsA, firstsiteind(ϕ, siteA))
+    C = noprime(C,tags="Site") * ϕdag[siteA]
+  end
+  for j in siteA+1:siteB-1
+    C = C * ϕ[j]
+    C = C * ϕdag[j]
+  end
+  if siteB == length(ϕ)
+    C = C * ϕ[siteB] * gate(obsB, firstsiteind(ϕ, siteB))
+    C = noprime(C,tags="Site") * ϕdag[siteB]
+  else
+    C = C * prime(ϕ[siteB],commonind(ϕ[siteB],ϕ[siteB+1])) * gate(obsB, firstsiteind(ϕ, siteB))
+    C = noprime(C,tags="Site") * ϕdag[siteB]
+  end
+  return real(C[])
+end
+
+measure(ψ::MPS, measurement::Tuple{String,Int,String,Int}) = 
+  measure(ψ, measurement, siteinds(ψ))
+
+function measure(ψ::MPS, measurement::Tuple{String,String}, s::Vector{<:Index})
+  N = length(ψ)
+  C = Matrix{Float64}(undef,N,N)
+  for siteA in 1:N
+    for siteB in 1:N
+      m = (measurement[1],siteA,measurement[2],siteB)
+      result = measure(ψ, m, s)
+      C[siteA,siteB] = result
+    end
+  end
+  return C
+end
+
+# for every sites
+measure(ψ::MPS, measurement::Tuple{String,String}) = 
+  measure(ψ::MPS, measurement, siteinds(ψ))
+
+
+
+
+"""
+    entanglemententropy(ψ::MPS; bond = nothing)
+
+Measure the entanglement entropy of an MPS `ψ` at `bond`.
+"""
+function entanglemententropy(ψ::MPS; bond = nothing)
+    # number of qubits
+    N = length(ψ)
+    
+    # make sure the state is normalized
+    normalize!(ψ)
+    
+    # if bond = nothing, take the center bond
+    bond = (isnothing(bond) ? N÷2 : bond)
+    @assert (bond < N)
+    
+    # gauge the MPS
+    orthogonalize!(ψ, bond)
```

Comment:
Remember `ψ = orthogonalize!(copy(ψ), bond)` to avoid modifying the input MPS. We can also define `orthogonalize(ψ, n) = orthogonalize!(copy(ψ), n)`, which should be in ITensors.jl if it is not already.